test/Handler/SitesSpec.hs

```diff
@@ -0,0 +1,90 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Handler.SitesSpec where
+
+import TestHelper hiding (createSite)
+import Model.Site
+import qualified Database.Persist as DB
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = withApp $ do
+    describe "SitesR" $ do
+        it "shows a list of sites for the correct user" $ do
+            user1 <- createUser "1"
+            user2 <- createUser "2"
+            runDB $ mapM_ (createSite $ entityKey user1) $
+                [ buildSite { siteName = "Site 1" }
+                , buildSite { siteName = "Site 2" }
+                ]
+
+            authenticateAs user1
+            get SitesR
+
+            htmlAnyContain "li" "Site 1"
+            htmlAnyContain "li" "Site 2"
+
+            authenticateAs user2
+            get SitesR
+
+            htmlNoneContain "li" "Site 1"
+            htmlNoneContain "li" "Site 2"
+
+        it "allows for creation of a site" $ do
+            user <- createUser "1"
+            authenticateAs user
+
+            get SitesR
```

Comment:
You need to visit the page containing the CSRF token to be able to post the
form. I could incorporate this into postForm though (especially if I rely
on the GET/POST being the same route).
On Mar 6, 2015 7:16 AM, "Joe Ferris" <notifications@github.com> wrote:

> In test/Handler/SitesSpec.hs
> <https://github.com/thoughtbot/carnival/pull/149#discussion_r25950763>:
>
> > +            get SitesR
> > +
> > +            htmlAnyContain "li" "Site 1"
> > +            htmlAnyContain "li" "Site 2"
> > +
> > +            authenticateAs user2
> > +            get SitesR
> > +
> > +            htmlNoneContain "li" "Site 1"
> > +            htmlNoneContain "li" "Site 2"
> > +
> > +        it "allows for creation of a site" $ do
> > +            user <- createUser "1"
> > +            authenticateAs user
> > +
> > +            get SitesR
>
> Is this step necessary? Does state persist between these fake requests?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/thoughtbot/carnival/pull/149/files#r25950763>.
>